### ISSUE TITLE
Tests are unreliable in kitty-items CI environment

### ISSUE THREAD

**Issue by 10thfloor** [onflow/flow-js-testing#47]
I noticed intermittent failures of tests in CI environment here: github.com/onflow/kitty-items
(Uncomment line #40 in `.github.workflows/integration-local.yml` and create a PR to observe the failure)

Tests run locally without failing using `npx jest` in `cadence/tests/js`

Things that did not fix the problem: 
- Increased `jest` timeout to 5 minutes
- Changed emulator port numbers to less popular location (8085 -> 7001 ...etc)
- Running jest tests sequentially (using `--runInBand` flag)

I suspect it has to do with how the tests are written. Starting and stopping the emulator in `beforeEach` blocks and recreating deployments every test. 

Suggestion: 
- Do setup and deployment once per suite, not per test.

**Comment by psiemens** [comment]
I think there's a chance that there's a port collision happening between emulator runs. Maybe we can use a method like this to dynamically assign a free port on emulator startup? https://davidwalsh.name/javascript-port-scanner

**Comment by samatechtw** [comment]
I think it's possible there are multiple contributors. I don't know if ports are an issue, I've tried incrementing and randomly assigning ports, and the rate of failure does not change.

For us, the failure is always:
```
    Response closed without headers

      at Object.onEnd (../../node_modules/flow-js-testing/node_modules/@onflow/sdk/dist/sdk.js:1:10175)
      at ../../node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js:1:24233
      at ../../node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js:1:11490
          at Array.forEach (<anonymous>)
      at e.rawOnError (../../node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js:1:11452)
      at e.onTransportEnd (../../node_modules/@improbable-eng/grpc-web/dist/grpc-web-client.js:1:10318)
      at ClientRequest.<anonymous> (../../node_modules/@improbable-eng/grpc-web-node-http-transport/lib/index.js:64:27)
```

I've tried a few things that have reduced the rate of failure, but it's not perfect yet (fingers crossed, CI has passed 5 times in a row at the time of writing).

For context, I'm running about 15 tests that are mostly modified from `kitty-items`. Before changes, the rate of failure was about 2 per 15 tests.

1. Rewrite `emulator.stop` to watch for process exit instead of using a timeout (`SIGKILL` is probably not necessary). This seems to reduce the failures to ~1/15 tests, but I didn't check it rigorously.
```js
// emulator.js
stop() {
    return new Promise((resolve, _reject) => {
      if (this.process) {
        this.process.on('exit', (code) => {
          this.process = null
          resolve(`Emulator stopped with code ${code}`)
        })
        this.process.kill('SIGKILL')
      }
    })
  }
```
2. Wait 3 seconds before starting each test (in `beforeEach`, before starting the emulator). This helped, but I still saw a lot of mid-test failures.
```
if (process.env.CI) {
  await new Promise((r) => setTimeout(r, 3000))
}
```
3. Wrap `flow-js-testing` call that results in grpc communication in a retry mechanism. This appears to work pretty well. I can write a PR, but I think it's a nasty hack and the real issue needs to be resolved in the SDK or emulator.
```
export const retryN = async (promise, n) => {
  let error = null
  for (let i = 0; i < n; i += 1) {
    try {
      return await promise
    } catch (e) {
      if (typeof e === 'string' && e.includes('Response closed without headers')) {
        console.log(`emulator cmd failed, retry ${n - (i + 1)} more times`)
        error = e
      } else {
        throw e
      }
    }
  }
  throw error
}
```

**Comment by brian-klktn** [comment]
I also run into the same problem and like to share my findings 
`flow-js-testing:0.1.13
mocha: 8.4.0
node: 14
emulator:v0.28.3
also running in github action
`

sample log of the failure. you can see line 2~4 are a successful emulator exit after test completion, but line `2021-09-30T03:02:10.1635489Z emulator exited with code null` shows the emulator failure right after `2021-09-30T03:02:10.1573960Z EMULATOR IS UP! Listening for events!` (0.00615 secs after)

```
2021-09-30T03:02:10.0073942Z     ✓ NFT service - Open an NFT (88ms)
2021-09-30T03:02:10.0091533Z LOG: time="2021-09-30T03:02:10Z" level=info msg="🛑  Server stopped"
2021-09-30T03:02:10.0092140Z 
2021-09-30T03:02:10.0092679Z emulator exited with code 0
2021-09-30T03:02:10.0575745Z 
2021-09-30T03:02:10.0577808Z   Order Service
2021-09-30T03:02:10.1217369Z LOG: time="2021-09-30T03:02:10Z" level=info msg="⚙️   Using service account 0xf8d6e0586b0a20c7" serviceAddress=f8d6e0586b0a20c7 serviceHashAlgo=SHA3_256 servicePrivKey=#################################################################### servicePubKey=####################################################################  serviceSigAlgo=ECDSA_P256
2021-09-30T03:02:10.1220835Z 
2021-09-30T03:02:10.1547018Z LOG: time="2021-09-30T03:02:10Z" level=info msg="📜  Flow contracts" FlowFees=0xe5a8b7f23e8b548f FlowServiceAccount=0xf8d6e0586b0a20c7 FlowStorageFees=0xf8d6e0586b0a20c7 FlowToken=0x0ae53cb6e3f42a79 FungibleToken=0xee82856bf20e2aa6
2021-09-30T03:02:10.1548596Z 
2021-09-30T03:02:10.1571411Z LOG: time="2021-09-30T03:02:10Z" level=info msg="🌱  Starting gRPC server on port 3509" port=3509
2021-09-30T03:02:10.1572680Z time="2021-09-30T03:02:10Z" level=info msg="🌱  Starting HTTP server on port 8020" port=8020
2021-09-30T03:02:10.1573320Z 
2021-09-30T03:02:10.1573960Z EMULATOR IS UP! Listening for events!
2021-09-30T03:02:10.1608775Z 
2021-09-30T03:02:10.1618767Z     1) "before all" hook: initWebServer for "Prep - #############################"
2021-09-30T03:02:10.1635489Z emulator exited with code null
2021-09-30T03:02:10.2125970Z 
```

**Comment by bluesign** [comment]
Hey, just saw the issue from @MaxStalker's mention. 

In a quick look this is a problem with Github Workers being slow. when you launch emulator till it is listening there is some time passing. (which is much bigger than local )

If you change test like this (in` beforeEach` and `afterEach`) ( I did 2000ms wait but I guess much less can be ok ) 

https://github.com/bluesign/kitty-items/blob/ddffc241cab7c321a7309c39609b745263409c3c/cadence/tests/js/test/kibble.test.js#L20

All workflow tests passes. 

https://github.com/bluesign/kitty-items/runs/4485889803?check_suite_focus=true

**Comment by MaxStalker** [comment]
@bluesign oh! that's quite the discovery! :exploding_head: 
I was actually suspecting conflicting ports due to the fact that `config()` would return a singleton :thinking: 
Which is why I solved this with `jest-serial-runner`:
https://github.com/onflow/flow-js-testing/issues/61

But I will check your solution as well, thank you @bluesign !

**Comment by samatechtw** [comment]
We started out with the same solution, but it still intermittently fails (the percentage of failures does go down as wait increases, though).

The long wait is somewhat inconvenient when running many tests, as well. Currently we solve it by restarting the emulator per test file, which comes with its own drawbacks.

**Comment by MaxStalker** [comment]
@samatechtw can you try that solution with `jest-serial-runner` outlines here:
https://github.com/onflow/flow-js-testing/issues/61

We made it work with KittyItems, but would be cool to get more real usage examples.
Thanks in advance! :bow:

**Comment by bluesign** [comment]
Yeah we can have something like: startEmulator waits for port availability, so it can wait only required time. ( it can try to run a script, when failed waits a second and try again etc)

**Comment by MaxStalker** [comment]
> Yeah we can have something like: startEmulator waits for port availability, so it can wait only required time. ( it can try to run a script, when failed waits a second and try again etc)

I've tried to run emulator on a new port every time - didn't help that much :)
Though maybe with this await solution it would work better, but it feels really hacky...

**Comment by samatechtw** [comment]
I'll try the `jest-serial-runner` as soon as I get the chance.

It's not just an issue with port availability, we tried:
- Waiting for port
- Waiting for the emulator process to fully close in `afterEach`

Both helped somewhat, but we still saw intermittent failures.

**Comment by brian-klktn** [comment]
My finding suggests that this is an out of resource issue. the Emulator immediately dies after spinning up with NULL as exit code. Indicates the suspect to be the kernel.

**Comment by bluesign** [comment]
> My finding suggests that this is an out of resource issue. the Emulator immediately dies after spinning up with NULL as exit code. Indicates the suspect to be the kernel.

I think it can be also port in use issue in that case

**Comment by kerrywei** [comment]
Here’s my investigation summary with KittyItem repo:
it is likely not caused by emulator being killed. reason being that we should be able to see logs with SIGKILL (or other signals), which is not the case (note: this does not rule out the possibility that emulator could be killed in other CI environment like @brian-klktn [point out](https://github.com/onflow/flow-js-testing/issues/47#issuecomment-931420404) )
it is likely not caused by emulator port taken, because: 1. if thats the case, the emulator should fail to start ; 2. emulators are successfully started and successfully stopped (each start is paired with a stop with exit code 0) by examining multiple failed test logs

### Root cause
It is extremely likely caused by js-testing-library relying on a [logging line](https://github.com/onflow/flow-js-testing/blob/master/src/emulator.js#L108) (from emulator) to determine whether the emulator is fully started, which is not very reliable. As you can see in that code block, the emulator logs “Starting HTTP server on port xxx” before the http/grpc servers are started; the js-testing-library is relying on this line to [resolve the promise](https://github.com/onflow/flow-js-testing/blob/master/src/emulator.js#L110) which is returned by the enulator.start() function; the Jest test framework would then assume the emulator is ready and hence kick off running the test logic. Hence there is quite a chance that the http or grpc servers are not really ready when the tests are executed, which explains that [“Response closed without headers” error](https://github.com/onflow/kitty-items/runs/4729884358?check_suite_focus=true).

### Temporary solution
#### Option A
Reduce the frequency of emulator start/stop, so that overall the test failure rate is lower. i.e. in Jest testing framework, you can consider using beforeAll and afterAll at the top level, instead of beforeEach and afterEach. Though this approach has drawbacks: 
- all tests in the same file are “sharing” the same storage layer — the emulator, which does not have a way to purge data, meaning your tests in the same file would need to be carefully constructed so that they do not pollute each other’s data.

#### Option B
Add delay in beforeEach and afterEach preparation call. A delay of 1 second suggests it is enough based on my tests — [25 test job executions are successful](https://github.com/onflow/kitty-items/actions/runs/1664079594/) after delay is introduced (as compared to previously 3 out of 6 jobs would fail without the delay). If this is still not enough, consider further increasing the delay

#### Option C
Add retry in logic that interact with the emulator blockchain, though this is cumbersome and would require you to revert these changes when the permanent solution below is out

### Permanent solution
Flow emulator: needs to provide a way to let other processes to check server readiness (https://github.com/onflow/flow-emulator/issues/111)
js-testing-library: needs to update  the logic to get status from the new interface from emulator instead of that logging (https://github.com/onflow/flow-js-testing/issues/72)
KittyItems: update code base to use the upgraded js-testing-library and remove the 1 second delay

**Comment by psiemens** [comment]
> Flow emulator: needs to provide a way to let other processes to check server readiness (onflow/flow-emulator#111)

@kerrywei [I left a comment on that issue](https://github.com/onflow/flow-emulator/issues/111#issuecomment-1007130169) with more details, but it should already be possible to ping the emulator and check if it's running.

Maybe that can be used to solve this?

**Comment by bluesign** [comment]
I think best solution is hidden in the problem.

> Starting HTTP server on port xxx 

Problem is this is not reliable. ( as starting can take unknown amount of time ) 

Maybe better to change it to: Started HTTP server on port xxxx after we started HTTP server. 

Other brute force solutions doesn't seem to elegant to me tbh.

**Comment by kerrywei** [comment]
Hello @bluesign , @samatechtw , and @brian-klktn , we just published a newer version of flow-js-testing -- 0.2.2-alpha.1, which should address the intermittent failre caused by testing lib prematurely flags the emulator is ready. It would be great to have you give it a shot and provide any feedback

**Comment by brian-klktn** [comment]
hi @kerrywei , thanks for the patch, we have seen improvements in our pipeline after the upgrade. thanks!

**Comment by kerrywei** [comment]
> hi @kerrywei , thanks for the patch, we have seen improvements in our pipeline after the upgrade. thanks!

Good to know that @brian-klktn . is there any number you can provide in terms of before/after behavior? 

we will close this issue soon given flow-js-testing library 0.2.2-alpha.3 is out -- we tested the pipeline 42 times, all passed without hicuups.

**Comment by kerrywei** [comment]
Hey folks, we've published 0.2.2 version of flow-js-testing lib which should address test failures caused by testing lib prematurely flags the emulator is ready. I'm closing this issue.
